cloud/linode/loadbalancers.go

```diff
@@ -229,6 +229,103 @@ func (l *loadbalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
 	return lbStatus, nil
 }
 
+// getNodeBalancerDeviceId gets the deviceID of the nodeBalancer that is attached to the firewall.
+func (l *loadbalancers) getNodeBalancerDeviceId(ctx context.Context, firewallID, nbID, page int) (int, bool, error) {
+	devices, err := l.client.ListFirewallDevices(ctx, firewallID, &linodego.ListOptions{PageSize: 500, PageOptions: &linodego.PageOptions{Page: page}})
+	if err != nil {
+		return 0, false, err
+	}
+
+	if len(devices) == 0 {
+		return 0, false, nil
+	}
+
+	for _, device := range devices {
+		if device.Entity.ID == nbID {
+			return device.ID, true, nil
+		}
+	}
+
+	return 0, false, nil
+}
+
+// updateNodeBalancerFirewall updates the firewall attached to the nodebalancer
+//
+// Firewall is updated in 3 scenario's:
+// - Add new firewall to the nodeBalancer
+// - Update firewall attached to the nodebalancer
+// - Remove firewall attached
+func (l *loadbalancers) updateNodeBalancerFirewall(ctx context.Context, service *v1.Service, nb *linodego.NodeBalancer) error {
+	var newFirewallID int
+	var err error
+
+	// get the new firewall id form the annotaiton (if any).
+	fwid, ok := service.GetAnnotations()[annLinodeCloudFirewallID]
+	if ok {
+		newFirewallID, err = strconv.Atoi(fwid)
+		if err != nil {
+			return err
+		}
+	}
+
+	// get the list of attached firewalls to the node-balancer.
+	firewalls, err := l.client.ListNodeBalancerFirewalls(ctx, nb.ID, &linodego.ListOptions{})
+	if err != nil {
+		if !ok {
+			if err.Error() != "[404] Not Found" {
```

Comment:
This is a bit icky, can we maybe use the `linodego.Error` type assertion and its `Code` field?